// react-globe.gl/dist/react-globe.gl.common.js
'use strict';

var fromKapsule = require('react-kapsule');
var GlobeKapsule = require('globe.gl');
var PropTypes = require('prop-types');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var fromKapsule__default = /*#__PURE__*/_interopDefaultLegacy(fromKapsule);
var GlobeKapsule__default = /*#__PURE__*/_interopDefaultLegacy(GlobeKapsule);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);

var GlobePropTypes = {
  width: PropTypes__default['default'].number,
  height: PropTypes__default['default'].number,
  backgroundColor: PropTypes__default['default'].string,
  backgroundImageUrl: PropTypes__default['default'].string,
  globeImageUrl: PropTypes__default['default'].string,
  bumpImageUrl: PropTypes__default['default'].string,
  showAtmosphere: PropTypes__default['default'].bool,
  showGraticules: PropTypes__default['default'].bool,
  onGlobeClick: PropTypes__default['default'].func,
  onGlobeRightClick: PropTypes__default['default'].func,
  pointsData: PropTypes__default['default'].arrayOf(PropTypes__default['default'].object),
  pointLat: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pointLng: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pointColor: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pointAltitude: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pointRadius: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pointResolution: PropTypes__default['default'].number,
  pointsMerge: PropTypes__default['default'].bool,
  pointsTransitionDuration: PropTypes__default['default'].number,
  pointLabel: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  onPointClick: PropTypes__default['default'].func,
  onPointRightClick: PropTypes__default['default'].func,
  onPointHover: PropTypes__default['default'].func,
  arcsData: PropTypes__default['default'].arrayOf(PropTypes__default['default'].object),
  arcStartLat: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcStartLng: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcEndLat: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcEndLng: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcColor: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].arrayOf(PropTypes__default['default'].string), PropTypes__default['default'].func]),
  arcAltitude: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcAltitudeAutoScale: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcStroke: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcCurveResolution: PropTypes__default['default'].number,
  arcCircularResolution: PropTypes__default['default'].number,
  arcDashLength: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcDashGap: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcDashInitialGap: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcDashAnimateTime: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  arcsTransitionDuration: PropTypes__default['default'].number,
  arcLabel: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  onArcClick: PropTypes__default['default'].func,
  onArcRightClick: PropTypes__default['default'].func,
  onArcHover: PropTypes__default['default'].func,
  polygonsData: PropTypes__default['default'].arrayOf(PropTypes__default['default'].object),
  polygonGeoJsonGeometry: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  polygonCapColor: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  polygonSideColor: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  polygonStrokeColor: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  polygonAltitude: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  polygonCapCurvatureResolution: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  polygonsTransitionDuration: PropTypes__default['default'].number,
  polygonLabel: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  onPolygonClick: PropTypes__default['default'].func,
  onPolygonRightClick: PropTypes__default['default'].func,
  onPolygonHover: PropTypes__default['default'].func,
  pathsData: PropTypes__default['default'].array,
  pathPoints: PropTypes__default['default'].oneOfType([PropTypes__default['default'].array, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pathPointLat: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pathPointLng: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pathPointAlt: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pathResolution: PropTypes__default['default'].number,
  pathColor: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].arrayOf(PropTypes__default['default'].string), PropTypes__default['default'].func]),
  pathStroke: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pathDashLength: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pathDashGap: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pathDashInitialGap: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pathDashAnimateTime: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  pathTransitionDuration: PropTypes__default['default'].number,
  pathLabel: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  onPathClick: PropTypes__default['default'].func,
  onPathRightClick: PropTypes__default['default'].func,
  onPathHover: PropTypes__default['default'].func,
  hexBinPointsData: PropTypes__default['default'].arrayOf(PropTypes__default['default'].object),
  hexBinPointLat: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  hexBinPointLng: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  hexBinPointWeight: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  hexBinResolution: PropTypes__default['default'].number,
  hexMargin: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].func]),
  hexTopColor: PropTypes__default['default'].func,
  hexSideColor: PropTypes__default['default'].func,
  hexAltitude: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].func]),
  hexTopCurvatureResolution: PropTypes__default['default'].number,
  hexBinMerge: PropTypes__default['default'].bool,
  hexTransitionDuration: PropTypes__default['default'].number,
  hexLabel: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  onHexClick: PropTypes__default['default'].func,
  onHexRightClick: PropTypes__default['default'].func,
  onHexHover: PropTypes__default['default'].func,
  hexPolygonsData: PropTypes__default['default'].arrayOf(PropTypes__default['default'].object),
  hexPolygonGeoJsonGeometry: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  hexPolygonColor: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  hexPolygonAltitude: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  hexPolygonResolution: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  hexPolygonMargin: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  hexPolygonCurvatureResolution: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  hexPolygonsTransitionDuration: PropTypes__default['default'].number,
  hexPolygonLabel: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  onHexPolygonClick: PropTypes__default['default'].func,
  onHexPolygonRightClick: PropTypes__default['default'].func,
  onHexPolygonHover: PropTypes__default['default'].func,
  labelsData: PropTypes__default['default'].arrayOf(PropTypes__default['default'].object),
  labelLat: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  labelLng: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  labelAltitude: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  labelRotation: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  labelText: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  labelSize: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  labelTypeFace: PropTypes__default['default'].object,
  labelColor: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  labelResolution: PropTypes__default['default'].number,
  labelIncludeDot: PropTypes__default['default'].oneOfType([PropTypes__default['default'].bool, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  labelDotRadius: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  labelDotOrientation: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  labelsTransitionDuration: PropTypes__default['default'].number,
  labelLabel: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  onLabelClick: PropTypes__default['default'].func,
  onLabelRightClick: PropTypes__default['default'].func,
  onLabelHover: PropTypes__default['default'].func,
  customLayerData: PropTypes__default['default'].arrayOf(PropTypes__default['default'].object),
  customThreeObject: PropTypes__default['default'].oneOfType([PropTypes__default['default'].object, PropTypes__default['default'].string, PropTypes__default['default'].func]),
  customThreeObjectUpdate: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  customLayerLabel: PropTypes__default['default'].oneOfType([PropTypes__default['default'].string, PropTypes__default['default'].func]),
  onCustomLayerClick: PropTypes__default['default'].func,
  onCustomLayerRightClick: PropTypes__default['default'].func,
  onCustomLayerHover: PropTypes__default['default'].func,
  enablePointerInteraction: PropTypes__default['default'].bool,
  pointerEventsFilter: PropTypes__default['default'].func,
  onZoom: PropTypes__default['default'].func
};

var Globe = fromKapsule__default['default'](GlobeKapsule__default['default'], {
  methodNames: [// bind methods
  'globeMaterial', 'pauseAnimation', 'resumeAnimation', 'pointOfView', 'scene', 'camera', 'renderer', 'postProcessingComposer', 'controls', 'getCoords', 'getScreenCoords', 'toGeoCoords', 'toGlobeCoords'],
  initPropNames: ['animateIn', 'waitForGlobeReady', 'rendererConfig']
});
Globe.displayName = 'Globe';
Globe.propTypes = GlobePropTypes;

module.exports = Globe;
